### PR TITLE
controller v1

### PR DESCRIPTION
**Pull request by reevefrancis17**
wrote controller state

### CODE REVIEW

File: firmware/Core/Src/Controller.c [hunk 1]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
+        if (avgTemp > 3) //on at 3C
+            actuate[1] = 1;
+        else if (avgTemp < 1)
+            actuate[1] = 0; //off at 1C
+        loopCounter =0;
+    }
+
+
+    loopCounter++;
+    return actuate;
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: states need to return the next state. This should remain "return CTRL_ACTUATE_FRIDGE;"

File: firmware/Core/Src/Controller.c [hunk 2]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
     return CTRL_DO_MATH;
 }
 
+float savedTemps[];
+float loopCounter;
+
+
 static Controller_State_t DoMath_State(void)
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c:110]: feel free to change the name of this state btw. Up to you, I just couldn't think of a great name.

File: firmware/Core/Src/Controller.c [hunk 3]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
     return CTRL_DO_MATH;
 }
 
+float savedTemps[];
+float loopCounter;
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: We should not have variables floating around like this. I suggest you make these static variables inside your state. that way they will preserve their values from call to call while still being local to your function.

File: firmware/Core/Src/Controller.c [hunk 4]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
+
+
 static Controller_State_t DoMath_State(void)
 {
-    return CTRL_ACTUATE_FRIDGE;
+    float temps[5] = {t1, t2, t3, t4, t5}; //5 temp probe readings
+    float temp_diffs[5]; //to store deltas between probe readings and avg temp
+    float avgTemp = 0; //mean temperature
+    float fan_treshold = 1; //desired fan treshold
+    int actuate[2] = {0, 0}; //first position is fan, second is compressor
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: Take a look at Actuators.h You'll find the definition to a struct. At the top of this (Controller.c) file, under the variables section, you'l find "static ActuatorCommands_t ActuatorCommands;". That's the instance of the struct you should populate. Once you do, you're done. The next state knows to read this variable to get the data you populated it with.


(Sidenote: I'll be moving the definition of that struct to Common.h, makes way more sense to have it there, but that doesen't affect you stuff in any way. Controller knows about that struct either way).

File: firmware/Core/Src/Controller.c [hunk 5]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
+float savedTemps[];
+float loopCounter;
+
+
 static Controller_State_t DoMath_State(void)
 {
-    return CTRL_ACTUATE_FRIDGE;
+    float temps[5] = {t1, t2, t3, t4, t5}; //5 temp probe readings
+    float temp_diffs[5]; //to store deltas between probe readings and avg temp
+    float avgTemp = 0; //mean temperature
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c:117]: Is this the temp we're shooting for? 0? I though Astra zenica was like 5 or 6.

File: firmware/Core/Src/Controller.c [hunk 6]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
+    float fan_treshold = 1; //desired fan treshold
+    int actuate[2] = {0, 0}; //first position is fan, second is compressor
+
+    for (int i=0, i<5; i++)
+        avgTemp += temps[i] / 5; //calculate mean temp
+
+    //fan control
+    if(loopCounter % 200 ==0) //every 2 seconds
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: Loop runs at CTRL_LOOP_FREQUENCY Hz, currently set to 50. 200 iterations will take 4 seconds.

File: firmware/Core/Src/Controller.c [hunk 7]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
 }
 
+float savedTemps[];
+float loopCounter;
+
+
 static Controller_State_t DoMath_State(void)
 {
-    return CTRL_ACTUATE_FRIDGE;
+    float temps[5] = {t1, t2, t3, t4, t5}; //5 temp probe readings
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: Go ahead and define a struct in Common.h that holds the temp readings. Then instantiate it as a static variable at the top of this file (same as with ActuatorCommands). The getData state will populate it for you, and you can read from it.

File: firmware/Core/Src/Controller.c [hunk 8]
```diff
@@ -103,9 +103,43 @@ static Controller_State_t LogData_State(void)
...
+    if(loopCounter % 200 ==0) //every 2 seconds
+        for (int i=0, i<5; i++)
+        {
+            temp_diffs[i] = temps[i] - avgTemp; //calculate deviation from mean temp.
+            if (temp_diffs[i] >= fan_treshold || temp_diffs[i] <= -1*fan_treshold)
+                actuate[0] = 1; //if any probe exceeds treshold, turn on internal fan
+        }
+
+    //compressor control
+    if(loopCounter % (50*120) ==0) //every 2 min
```
Comment (Fberbari) [inline firmware/Core/Src/Controller.c]: This isn't really ideal. This forces you to only make a decision every 2 minutes, but remember you can turn the compressor off any time. After that, you need to wait 75s from the time it was turned off before turning it back on.

**Review by Fberbari — CHANGES_REQUESTED** [review]
Core stuff seems to be there. Few fixes though. Thanks!

File: firmware/Core/Src/Controller.c (hunk 5 above)
Comment (reevefrancis17) [reply firmware/Core/Src/Controller.c:117]: I just initialized it to 0 but it will maintain between 1C and 3C, in our abstract I think we said 2C

**Review by reevefrancis17 — COMMENTED** [review]
made the requested changes

**Pull request closed without merging**